### PR TITLE
Charts for survey responses and admin dashboard

### PR DESCRIPTION
**Pull request by FatimaMubeenSh**
Survey's response charts: home/survey_id/survey_charts 
Admin dashboard charts: home/main_charts

### CODE REVIEW

File: app/assets/javascripts/survey_charts.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+$(document).ready(function() {
+  $('.survey-chart').find('canvas').each(function(){
```
Comment (zohaibbutt13) [inline app/assets/javascripts/survey_charts.js]: .js-survey-chart

File: app/controllers/home_controller.rb [hunk 2]
```diff
@@ -18,14 +18,26 @@ def packages
   # get home/display_surveys
   def display_surveys
     if params[:filter_by] == 'expiry'
-      @surveys = Survey.get_expired_surveys
+      @surveys = Survey.expired_surveys
```
Comment (zohaibbutt13) [inline app/controllers/home_controller.rb]: Survey.expired_surveys -> @current_company.surveys.expired_surveys

File: app/models/question.rb [hunk 3]
```diff
@@ -2,13 +2,39 @@ class Question < ActiveRecord::Base
   belongs_to :survey, inverse_of: :questions
   has_many :options, dependent: :destroy, inverse_of: :question, autosave: true
   accepts_nested_attributes_for :options
-  has_many :answers, dependent: :destroy
+  has_many :answers
```
Comment (zohaibbutt13) [inline app/models/question.rb]: why removed dependent: :destroy

File: app/models/question.rb [hunk 4]
```diff
@@ -2,13 +2,39 @@ class Question < ActiveRecord::Base
...
   belongs_to :company
 
   validates :statement, presence: true, length: { maximum: 500 }
 
   before_save :mark_option_for_removal
 
+  def mark_option_for_removal
+    options.each do |option|
+      option.mark_for_destruction if option.detail == 'nill'
```
Comment (zohaibbutt13) [inline app/models/question.rb]: why not .nil?

File: app/models/question.rb [hunk 5]
```diff
@@ -2,13 +2,39 @@ class Question < ActiveRecord::Base
...
+    end
+  end
+
+  def answer_stats
+    stats_data = []
+    options.each do |option|
+      option_count = 0
+      answers.each do |answer|
+        option_count += 1 if answer.option_id == option.id
```
Comment (zohaibbutt13) [inline app/models/question.rb]: can we use this by using single query ?

File: app/models/question.rb [hunk 6]
```diff
@@ -2,13 +2,39 @@ class Question < ActiveRecord::Base
...
+      end
+      stats_data.push(option_count)
+    end
+    stats_data
+  end
+
+  def options_labels
+    labels = []
+    options.each do |option|
```
Comment (zohaibbutt13) [inline app/models/question.rb]: single query

File: app/controllers/home_controller.rb [hunk 7]
```diff
@@ -14,4 +14,37 @@ def packages
...
+    respond_to do |format|
+      format.html
+    end
+  end
+
+  def main_charts
+    @surveys_stats = [@current_company.surveys.expired_surveys.count,
+                      @current_company.surveys.active_surveys.count]
+    @surveys_stats_labels = ['Expired', 'Active']
+    @latest_surveys_labels = @current_company.surveys.latest_surveys.limit(3).pluck(:name)
```
Comment (zohaibbutt13) [inline app/controllers/home_controller.rb]: active_surveys check

File: app/controllers/home_controller.rb [hunk 8]
```diff
@@ -14,4 +14,37 @@ def packages
...
+    @latest_surveys_responses_count = @current_company.surveys.latest_surveys_responses(3)
+    respond_to do |format|
+      format.html
+    end
+  end
+
+  # get home/:id/survey_charts
+  def survey_charts
+    @survey = Survey.find(params[:id])
```
Comment (zohaibbutt13) [inline app/controllers/home_controller.rb]: cancancan will provide @survey

File: app/models/question.rb [hunk 9]
```diff
@@ -9,6 +9,24 @@ class Question < ActiveRecord::Base
...
 
+  def mark_option_for_removal
+    options.each do |option|
+      option.mark_for_destruction if option.detail.nil?
+    end
+  end
+
+  def answer_stats
+    stats_data = []
+    options.each do |option|
```
Comment (zohaibbutt13) [inline app/models/question.rb]: do it in single query

File: app/models/survey.rb [hunk 10]
```diff
@@ -20,26 +20,32 @@ class Survey < ActiveRecord::Base
...
-  def self.get_active_surveys
-    Survey.active_surveys
+  # Returns an array having surveys respones count
+  def self.latest_surveys_responses(count)
+    survey_responses = []
+    latest_surveys.limit(count)
+                  .joins('LEFT OUTER JOIN user_responses on surveys.id = user_responses.survey_id')
+                  .select('surveys.id, COUNT(user_responses.id) AS responses_count')
+                  .group('surveys.id')
+                  .each do |survey|
```
Comment (zohaibbutt13) [inline app/models/survey.rb]: why loop after join ?

File: app/models/survey.rb [hunk 11]
```diff
@@ -20,26 +20,32 @@ class Survey < ActiveRecord::Base
...
+                    survey_responses.push(survey.responses_count)
+                  end
+    survey_responses
   end
 
   def mark_question_for_removal
```
Comment (zohaibbutt13) [inline app/models/survey.rb]: use of this method?

File: app/models/survey.rb [hunk 12]
```diff
@@ -20,26 +20,32 @@ class Survey < ActiveRecord::Base
...
     questions.each do |question|
       question.mark_for_destruction if question.statement == '-1'
     end
   end
-  
+
   def self.save?(survey)
```
Comment (zohaibbutt13) [inline app/models/survey.rb]: why self.save ?

File: app/views/home/display_surveys.html.erb [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+    <a class="btn btn-success my-auto" href="#_">Add Survey</a>
+  </div>
+  <div class="row d-flex justify-content-start">
+    <p class="text-muted">Show Surveys: </p>
+    <div class="dropdown mx-2 mb-2">
+      <button class="btn btn-secondary dropdown-toggle" type="button" id="dropdownMenuButton" data-toggle="dropdown">
+        Select
+      </button>
+      <div class="dropdown-menu" aria-labelledby="dropdownMenuButton">
+        <%= link_to "Expired", {controller: "home", action: "display_surveys", filter_by: 'expiry' }, class: "dropdown-item" %>
```
Comment (zohaibbutt13) [inline app/views/home/display_surveys.html.erb]: can we use rails path here ?

File: app/views/home/display_surveys.html.erb [hunk 14]
```diff
@@ -0,0 +1,47 @@
+<div class="container">
+  <div class="row d-flex justify-content-between">
+    <h1>Surveys</h1>
+    <a class="btn btn-success my-auto" href="#_">Add Survey</a>
```
Comment (zohaibbutt13) [inline app/views/home/display_surveys.html.erb]: ability and link path

File: app/views/layouts/_activity_history.html.erb [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+      <tr>
+        <th scope="col">#</th>
+        <th scope="col">Activity</th>
+        <th scope="col">Time</th>
+      </tr>
+    </thead>
+    <tbody>
+      <% @activities.each_with_index do |activity, index| %>
+        <tr>
+          <th scope="row"><%= index + 1%></th>
```
Comment (zohaibbutt13) [inline app/views/layouts/_activity_history.html.erb]: th in tbody !!!

File: app/views/layouts/_activity_history.html.erb [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+  <table class="table table-bordered table-striped mb-0">
+    <thead>
+      <tr>
+        <th scope="col">#</th>
+        <th scope="col">Activity</th>
+        <th scope="col">Time</th>
+      </tr>
+    </thead>
+    <tbody>
+      <% @activities.each_with_index do |activity, index| %>
```
Comment (zohaibbutt13) [inline app/views/layouts/_activity_history.html.erb]: if you need starting index from 1 then use each.with_index(1)

File: app/views/surveys/survey_charts.html.erb [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+      <h3>Question: <%= question.statement %></h3>
+      <table class="table table-striped">
+        <thead class="thead-dark">
+          <tr>
+            <th scope="col">#</th>
+            <th scope="col">Responses</th>
+          </tr>
+        </thead>
+        <tbody>
+          <% question.answers.each_with_index do |answer, index| %>
```
Comment (zohaibbutt13) [inline app/views/surveys/survey_charts.html.erb]: each.with_index(1)

Comment (FatimaMubeenSh) [reply app/views/surveys/survey_charts.html.erb]: done

File: app/views/layouts/_activity_history.html.erb (hunk 16 above)
Comment (FatimaMubeenSh) [reply app/views/layouts/_activity_history.html.erb]: done

File: app/views/layouts/_activity_history.html.erb (hunk 15 above)
Comment (FatimaMubeenSh) [reply app/views/layouts/_activity_history.html.erb]: resolved

File: app/views/home/display_surveys.html.erb (hunk 14 above)
Comment (FatimaMubeenSh) [reply app/views/home/display_surveys.html.erb]: file has been removed

File: app/views/companies/dashboard.html.erb [hunk 18]
```diff
@@ -1,18 +1,20 @@
...
+  <% if @user_setting.show_graphs? %>
+    <div id="graphs" class="my-2">
+      <h1>Graphs</h1>
+      <%= render 'layouts/main_charts', locals: { stats_data: @surveys_stats, stats_labels: @surveys_stats_labels, latest_surveys_labels: @latest_surveys_labels, latest_surveys: @latest_surveys_responses_count } %>
+    </div>
+  <% end %>
+  <% if @user_setting.show_history? %>
+    <div id="histroy" class="my-2">
+      <h1 >History</h1>
+      <%= render 'layouts/activity_history', locals: { user_activity: @activities } %>
```
Comment (zohaibbutt13) [inline app/views/companies/dashboard.html.erb]: user_activity -> user_activities 
use render partial: if you are rendering a html page

File: app/controllers/surveys_controller.rb [hunk 19]
```diff
@@ -110,4 +110,12 @@ def destroy
       render action: :show
     end
   end
+
+  # get survey/:id/survey_charts
+  def survey_charts
+    @survey = Survey.find(params[:id])
```
Comment (zohaibbutt13) [inline app/controllers/surveys_controller.rb]: find through current_company and include other resources you need

File: config/database.yml [hunk 20]
```diff
@@ -10,7 +10,7 @@ development:
   encoding: utf8
   database: survey_system
   username: root
-  password: test
+  password: MyDb97.11
```
Comment (zohaibbutt13) [inline config/database.yml]: revert this file

File: app/controllers/companies_controller.rb [hunk 21]
```diff
@@ -6,6 +6,15 @@ class CompaniesController < ApplicationController
   end
 
   def dashboard
+    @activities = @current_company.activities.get_user_activities(current_user)
+    @surveys_stats = [@current_company.surveys.expired_surveys.count,
+                      @current_company.surveys.active_surveys.count]
+    @surveys_stats_labels = ['Expired', 'Active']
```
Comment (zohaibbutt13) [inline app/controllers/companies_controller.rb]: create constant

File: app/views/companies/dashboard.html.erb [hunk 22]
```diff
@@ -1,18 +1,20 @@
...
-    <% end %>
-    <% if @user_setting.show_history? %>
-      <div id="histroy">
-        <h1 >History</h1>
-      </div>
-    <% end %>
+  <div class="col-lg-9">
+  <% if @user_setting.show_graphs? %>
+    <div id="graphs" class="my-2">
+      <h1>Graphs</h1>
```
Comment (zohaibbutt13) [inline app/views/companies/dashboard.html.erb]: move to en.yml (all hard coded messages)

Comment (FatimaMubeenSh) [reply app/views/companies/dashboard.html.erb]: This is just a heading? Should we move it too?

**Pull request merged**